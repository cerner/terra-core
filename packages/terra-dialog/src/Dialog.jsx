import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import Button from 'terra-button';
import ContentContainer from 'terra-content-container';
import styles from './Dialog.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * React node to be placed within the header area of the dialog.
   */
  header: PropTypes.node.isRequired,
  /**
   * React node to be placed within the footer area of the dialog.
   * Content exceeding the available space will be hidden.
   */
  footer: PropTypes.node.isRequired,
  /**
   * The children to be placed within the main content area of the dialog.
   */
  children: PropTypes.node,
  /**
   * Callback function for when the close button is clicked. The close button will not display if this is not set.
   * On small viewports a back button will be displayed instead of a close button.
   */
  onClose: PropTypes.func,
};

const defaultProps = {
  onClose: null,
  children: null,
};

const contextTypes = {
  /* eslint-disable consistent-return */
  intl: (context) => {
    if (context.intl === undefined) {
      return new Error('Component is internationalized, and must be wrapped in terra-base');
    }
  },
};

const Dialog = ({
  header, footer, onClose, children, ...customProps
}, { intl }) => {
  const dialogClassNames = cx([
    'dialog',
    customProps.className,
  ]);

  const closeText = intl.formatMessage({ id: 'Terra.dialog.close' });
  const closeButton = onClose ? <div className={cx('dialog-header-close')}><Button variant="utility" text={closeText} onClick={onClose} isIconOnly icon={<span className={cx('close-icon')} />} /></div> : null;

  const dialogHeader = (
    <div className={cx('dialog-header')}>
      <div className={cx('dialog-header-title')}>{header}</div>
      {closeButton}
    </div>
  );

  return (
    <div {...customProps} className={dialogClassNames}>
      <ContentContainer fill header={dialogHeader} footer={<div className={cx('dialog-footer')}>{footer}</div>}>
        <div className={cx('dialog-body')}>{children}</div>
      </ContentContainer>
    </div>
  );
};

Dialog.propTypes = propTypes;
Dialog.defaultProps = defaultProps;
Dialog.contextTypes = contextTypes;

export default Dialog;
