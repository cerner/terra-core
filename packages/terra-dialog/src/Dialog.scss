:local {
  .dialog {
    height: 100%;
    width: 100%;
  }

  .dialog-header {
    background-color: var(--terra-dialog-header-background-color, transparent);
    border-bottom: var(--terra-dialog-header-bottom-border, none);
    color: var(--terra-dialog-header-foreground-color, #1c1f21);
  }

  .dialog-body {
    background-color: var(--terra-dialog-body-background-color, transparent);
    color: var(--terra-dialog-body-foreground-color, #1c1f21);
    padding: var(--terra-dialog-body-padding, 0.7143rem);
  }

  .dialog-footer {
    background-color: var(--terra-dialog-footer-background-color, #f4f4f4);
    border-top: var(--terra-dialog-footer-bottom-top, 1px solid #dedfe0);
    color: var(--terra-dialog-footer-foreground-color, #1c1f21);
    padding: var(--terra-dialog-footer-padding, 0.7143rem);
  }
}
