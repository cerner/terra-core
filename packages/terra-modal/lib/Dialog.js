'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _terraButton = require('terra-button');

var _terraButton2 = _interopRequireDefault(_terraButton);

var _IconClose = require('terra-icon/lib/icon/IconClose');

var _IconClose2 = _interopRequireDefault(_IconClose);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * The <Dialog/> component contains the information in a modal, including header/body/footer.
 * It supports scrolling of body content and fixed footer at the bottom if the body overflows.
 * It supports full-screen dialog and sizes with max-width at different breakpoints.
 */
var propTypes = {
  /**
   * Set aria-label for modal dialog
   */
  ariaLabel: _react.PropTypes.string.isRequired,
  /**
   * Set modal dialog to full screen
   */
  isFullScreen: _react.PropTypes.bool,
  /**
   * Set the max-width for modal dialog
   */
  size: _react.PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
  /**
   * Set the height of modal dialog in px, max-height = windows.height * 85
   */
  height: _react.PropTypes.number,
  /**
   * Callback function for closing the modal
   */
  onRequestClose: _react.PropTypes.func.isRequired,
  /**
   * Displays close button on the top right(ltr) of container
   */
  hasCloseButton: _react.PropTypes.bool,
  /**
   * Modal header
   */
  header: _react.PropTypes.element,
  /**
  * Modal body
  */
  body: _react.PropTypes.element,
  /**
  * Modal footer
  */
  footer: _react.PropTypes.element
};

var defaultProps = {
  isFullScreen: false,
  size: 'medium',
  hasCloseButton: true
};

var Dialog = function (_React$Component) {
  (0, _inherits3.default)(Dialog, _React$Component);

  function Dialog() {
    (0, _classCallCheck3.default)(this, Dialog);

    var _this = (0, _possibleConstructorReturn3.default)(this, (Dialog.__proto__ || (0, _getPrototypeOf2.default)(Dialog)).call(this));

    _this.state = {};
    _this.updateBodyHeight = _this.updateBodyHeight.bind(_this);
    return _this;
  }

  (0, _createClass3.default)(Dialog, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.updateBodyHeight();
      window.addEventListener('resize', this.updateBodyHeight);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this.updateBodyHeight);
    }
  }, {
    key: 'updateBodyHeight',
    value: function updateBodyHeight() {
      var headerHeight = this.dialog.firstElementChild.offsetHeight;
      var footerHeight = this.dialog.lastElementChild.offsetHeight;
      var headerfooter = headerHeight + footerHeight;
      if (this.props.isFullScreen) {
        this.setState({
          bodyHeight: window.innerHeight - headerfooter,
          bodyMaxHeight: undefined
        });
      } else if (this.props.height) {
        var height = Math.min(this.props.height - headerfooter, window.innerHeight * 0.85 - headerfooter);
        this.setState({
          bodyHeight: height,
          bodyMaxHeight: undefined
        });
      } else {
        this.setState({
          bodyMaxHeight: window.innerHeight * 0.85 - headerfooter,
          bodyHeight: undefined
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var bodyStyle = {
        maxHeight: this.state.bodyMaxHeight,
        height: this.state.bodyHeight
      };
      var dialogClass = (0, _classnames2.default)('terra-Modal-dialog', 'terra-Modal--' + this.props.size, { 'terra-Modal--fullScreen': this.props.isFullScreen });
      return _react2.default.createElement(
        'div',
        {
          role: 'dialog',
          tabIndex: -1,
          'aria-label': this.props.ariaLabel,
          className: dialogClass,
          ref: function ref(dialog) {
            _this2.dialog = dialog;
          }
        },
        _react2.default.createElement(
          'div',
          { className: 'terra-Modal-header' },
          _react2.default.createElement(
            'div',
            null,
            this.props.header
          ),
          this.props.hasCloseButton && _react2.default.createElement(_terraButton2.default, {
            'aria-label': 'modal-close',
            title: 'modal-close',
            className: 'terra-Modal--close',
            icon: _react2.default.createElement(_IconClose2.default, null),
            onClick: this.props.onRequestClose
          })
        ),
        _react2.default.createElement(
          'div',
          { className: 'terra-Modal-body', style: bodyStyle },
          this.props.body
        ),
        _react2.default.createElement(
          'div',
          null,
          this.props.footer
        )
      );
    }
  }]);
  return Dialog;
}(_react2.default.Component);

Dialog.propTypes = propTypes;
Dialog.defaultProps = defaultProps;

exports.default = Dialog;