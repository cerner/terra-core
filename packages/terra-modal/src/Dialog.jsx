import React, { PropTypes } from 'react';
import classNames from 'classnames';
import Button from 'terra-button';
import CloseButton from 'terra-icon/lib/icon/IconClose';

/**
 * The <Dialog/> component contains the information in a modal, including header/body/footer.
 * It supports scrolling of body content and fixed footer at the bottom if the body overflows.
 * It supports full-screen dialog and sizes with max-width at different breakpoints.
 */
const propTypes = {
  /**
   * Set aria-label for modal dialog
   */
  ariaLabel: PropTypes.string.isRequired,
  /**
   * Set modal dialog to full screen
   */
  isFullScreen: PropTypes.bool,
  /**
   * Set the max-width for modal dialog
   */
  size: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
  /**
   * Set the height of modal dialog in px, max-height = windows.height * 85
   */
  height: PropTypes.number,
  /**
   * Callback function for closing the modal
   */
  onRequestClose: PropTypes.func.isRequired,
  /**
   * Displays close button on the top right(ltr) of container
   */
  hasCloseButton: PropTypes.bool,
  /**
   * Modal header
   */
  header: PropTypes.element,
  /**
  * Modal body
  */
  body: PropTypes.element,
  /**
  * Modal footer
  */
  footer: PropTypes.element,
};


const defaultProps = {
  isFullScreen: false,
  size: 'medium',
  hasCloseButton: true,
};

class Dialog extends React.Component {
  constructor() {
    super();
    this.state = {};
    this.updateBodyHeight = this.updateBodyHeight.bind(this);
  }

  componentDidMount() {
    this.updateBodyHeight();
    window.addEventListener('resize', this.updateBodyHeight);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updateBodyHeight);
  }

  updateBodyHeight() {
    const headerHeight = this.dialog.firstElementChild.offsetHeight;
    const footerHeight = this.dialog.lastElementChild.offsetHeight;
    const headerfooter = headerHeight + footerHeight;
    if (this.props.isFullScreen) {
      this.setState({
        bodyHeight: window.innerHeight - headerfooter,
        bodyMaxHeight: undefined,
      });
    } else if (this.props.height) {
      const height = Math.min(
        this.props.height - headerfooter,
        (window.innerHeight * 0.85) - headerfooter,
      );
      this.setState({
        bodyHeight: height,
        bodyMaxHeight: undefined,
      });
    } else {
      this.setState({
        bodyMaxHeight: (window.innerHeight * 0.85) - headerfooter,
        bodyHeight: undefined,
      });
    }
  }

  render() {
    const bodyStyle = {
      maxHeight: this.state.bodyMaxHeight,
      height: this.state.bodyHeight,
    };
    const dialogClass = classNames(
      'terra-Modal-dialog',
      `terra-Modal--${this.props.size}`,
      { 'terra-Modal--fullScreen': this.props.isFullScreen },
    );
    return (
      <div
        role="dialog"
        tabIndex={-1}
        aria-label={this.props.ariaLabel}
        className={dialogClass}
        ref={(dialog) => { this.dialog = dialog; }}
      >
        <div className="terra-Modal-header">
          <div>{this.props.header}</div>
          {
            this.props.hasCloseButton &&
            <Button
              aria-label="modal-close"
              title="modal-close"
              className="terra-Modal--close"
              icon={<CloseButton />}
              onClick={this.props.onRequestClose}
            />
          }
        </div>
        <div className="terra-Modal-body" style={bodyStyle}>
          {this.props.body}
        </div>
        <div>
          {this.props.footer}
        </div>
      </div>
    );
  }
}

Dialog.propTypes = propTypes;
Dialog.defaultProps = defaultProps;

export default Dialog;
