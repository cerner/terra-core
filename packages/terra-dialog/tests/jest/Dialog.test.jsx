import React from 'react';
/* eslint-disable-next-line import/no-extraneous-dependencies */
import { shallowWithIntl } from 'terra-enzyme-intl';
import Dialog from '../../src/Dialog';

describe('Dialog', () => {
  const defaultRender = <Dialog header="Header Content" footer="Footer Content">some body content</Dialog>;

  // Snapshot Test
  it('should render a default component', () => {
    const wrapper = shallowWithIntl(defaultRender);
    expect(wrapper).toMatchSnapshot();
  });

  // Structure Tests
  it('should have the class dialog', () => {
    const wrapper = shallowWithIntl(defaultRender);
    expect(wrapper.prop('className')).toContain('dialog');
  });

  it('should render a Dialog with merged attributes', () => {
    const wrapper = shallowWithIntl(<Dialog header="Header Content" footer="Footer Content" className="TestClass">some body content</Dialog>);
    expect(wrapper).toMatchSnapshot();
  });
});

it('throws error on missing locale prop in Base', () => {
  global.console = { error: jest.fn() };

  expect(() => {
    render(<Dialog header="Header Content" footer="Footer Content" />);
  }).toThrowError();
  // eslint-disable-next-line no-console
  expect(console.error).toBeCalledWith(expect.stringContaining('Component is internationalized, and must be wrapped in terra-base'));
});
