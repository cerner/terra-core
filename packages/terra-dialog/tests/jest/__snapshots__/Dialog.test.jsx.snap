// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should render a Dialog with merged attributes 1`] = `
<div
  className="dialog TestClass"
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
>
  <ContentContainer
    fill={true}
    footer={
      <div
        className="dialog-footer"
      >
        Footer Content
      </div>
    }
    header={
      <div
        className="dialog-header"
      >
        <div
          className="dialog-header-title"
        >
          Header Content
        </div>
      </div>
    }
  >
    <div
      className="dialog-body"
    >
      some body content
    </div>
  </ContentContainer>
</div>
`;

exports[`Dialog should render a default component 1`] = `
<div
  className="dialog"
  intl={
    Object {
      "defaultFormats": Object {},
      "defaultLocale": "en",
      "formatDate": [Function],
      "formatHTMLMessage": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelative": [Function],
      "formatTime": [Function],
      "formats": Object {},
      "formatters": Object {
        "getDateTimeFormat": [Function],
        "getMessageFormat": [Function],
        "getNumberFormat": [Function],
        "getPluralFormat": [Function],
        "getRelativeFormat": [Function],
      },
      "locale": "en",
      "messages": null,
      "now": [Function],
      "onError": [Function],
      "textComponent": "span",
      "timeZone": null,
    }
  }
>
  <ContentContainer
    fill={true}
    footer={
      <div
        className="dialog-footer"
      >
        Footer Content
      </div>
    }
    header={
      <div
        className="dialog-header"
      >
        <div
          className="dialog-header-title"
        >
          Header Content
        </div>
      </div>
    }
  >
    <div
      className="dialog-body"
    >
      some body content
    </div>
  </ContentContainer>
</div>
`;
